### PR TITLE
Implementation of labeling engine rules

### PR DESCRIPTION
**Pull request by nyalldawson**
See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/299

Implements the API framework for setting advanced labeling engine rules for a project, and implements 4 initial rule types:

- QgsLabelingEngineRuleMinimumDistanceLabelToFeature: prevents labels   being placed too *close* to features from a different layer
- QgsLabelingEngineRuleMaximumDistanceLabelToFeature: prevents labels   being placed too *far* from features from a different layer
- QgsLabelingEngineRuleMinimumDistanceLabelToLabel: prevents labels   being placed too close to labels from a different layer
- QgsLabelingEngineRuleAvoidLabelOverlapWithFeature: prevents labels   being placed overlapping features from a different layer

(note that the first 3 rules require a build based on GEOS >= 3.10, they are not available for older GEOS builds)

Also implements a registry for storing available rule classes, and serialization of rules and configuration in QGIS projects

Now with GUI:

https://github.com/user-attachments/assets/a8854ce4-4a55-4c2f-abf9-8cc68517ceae

Sponsored by Rubicon Concierge Real Estate

### CODE REVIEW

**Comment by nyalldawson** [comment]
Now with GUI, this is ready for review...

**Comment by DelazJ** [comment]
Nice @nyalldawson! What about showing involved layers next to the rule in the "Labeling rules" list? It might help to quickly identify the rule you want to update...

**Comment by nyalldawson** [comment]
@DelazJ you can set your own name for rules, which is used in that list. I was thinking that could be better for rule identification.

Maybe we could show a summary including involved layers in a tooltip for the list item?

**Comment by DelazJ** [comment]
Ah, great! I only watched the screencast which didn't showcase that.
> Maybe we could show a summary including involved layers in a tooltip for the list item?

+1 from me. When you are out of inspiration for rules naming, this could be handful 😄

**Comment by nyalldawson** [comment]
@DelazJ 
here you go:

https://github.com/user-attachments/assets/a22f3f06-4110-47fe-bdf8-5bd4c32a22c1

File: src/core/labeling/rules/qgslabelingenginerule.h [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    /**
+     * Returns a user-friendly description of the rule.
+     *
+     * This should include the rule name() if set, and other useful details for users
+     * to quickly identify the rule's purpose when shown in a tooltip.
+     *
+     * The returned string may contain HTML formatting.
+     */
+    virtual QString description() const;
```
Comment (nirvn) [inline src/core/labeling/rules/qgslabelingenginerule.h:189]: An icon representation could be nice too (not blocking this PR for that, just food for thoughts).

File: src/core/labeling/rules/qgslabelingenginerule.h [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+     * The default implementation returns the same bounds.
+     */
+    virtual QgsRectangle modifyCandidateConflictSearchBoundingBox( const QgsRectangle &candidateBounds ) const SIP_SKIP;
+
+    /**
+     * Returns TRUE if a labeling \a candidate violates the rule and should be eliminated.
+     *
+     * The default implementation returns FALSE.
+     */
+    virtual bool candidateIsIllegal( const pal::LabelPosition *candidate, QgsLabelingEngineContext &context ) const SIP_SKIP;
```
Comment (nirvn) [inline src/core/labeling/rules/qgslabelingenginerule.h:276]: I'm not sure why but I don't like the word 'illegal' here. What about candidateIsForbidden, candiateIsProhibited, candidateIsDisallowed, candidateIsBanned ? We're breaching rules not laws :wink:

File: src/core/labeling/rules/qgslabelingenginerule_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+
+#include "qgslabelingenginerule_impl.h"
+#include "qgsunittypes.h"
+#include "qgssymbollayerutils.h"
+#include "labelposition.h"
+#include "feature.h"
+#include "qgsvectorlayerfeatureiterator.h"
+#include "qgsthreadingutils.h"
+#include "qgsspatialindex.h"
+#include "qgsgeos.h"
```
Comment (nirvn) [inline src/core/labeling/rules/qgslabelingenginerule_impl.cpp:24]:
```suggestion
#include "qgslabelingenginerule_impl.h"
#include "qgsunittypes.h"
#include "qgssymbollayerutils.h"
#include "qgsvectorlayerfeatureiterator.h"
#include "qgsthreadingutils.h"
#include "qgsspatialindex.h"
#include "qgsgeos.h"
#include "labelposition.h"
#include "feature.h"
```

File: src/core/labeling/rules/qgslabelingenginerule_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+}
+
+QString QgsLabelingEngineRuleMaximumDistanceLabelToFeature::id() const
+{
+  return QStringLiteral( "maximumDistanceLabelToFeature" );
+}
+
+QString QgsLabelingEngineRuleMaximumDistanceLabelToFeature::displayType() const
+{
+  return QObject::tr( "Pull Labels toward Features" );
```
Comment (nirvn) [inline src/core/labeling/rules/qgslabelingenginerule_impl.cpp:291]: I'm pretty sure one must capitalize toward(s) in a title case sentence.

```suggestion
  return QObject::tr( "Pull Labels Toward Features" );
```

File: src/core/labeling/rules/qgslabelingengineruleregistry.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include "qgslabelingenginerule_impl.h"
+#include <geos_c.h>
+
+QgsLabelingEngineRuleRegistry::QgsLabelingEngineRuleRegistry()
+{
+#if GEOS_VERSION_MAJOR>3 || ( GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR>=10 )
+  addRule( new QgsLabelingEngineRuleMinimumDistanceLabelToFeature() );
+  addRule( new QgsLabelingEngineRuleMaximumDistanceLabelToFeature() );
+  addRule( new QgsLabelingEngineRuleMinimumDistanceLabelToLabel() );
+#endif
```
Comment (nirvn) [inline src/core/labeling/rules/qgslabelingengineruleregistry.cpp:26]: If we're conditionally adding rules to the registry based on GEOS version, that'll mean a project shared across multiple users with different GEOS versions (or projects processed on a server) will result in rule loss. Is that something you've thought about and are we OK with that? 

An alternative would be to add a QLabel note in the rule widget UI saying 'this rule will be ignored on your system' or something.

File: src/core/pal/pal.cpp [hunk 6]
```diff
@@ -396,17 +428,18 @@ std::unique_ptr<Problem> Pal::extractProblem( const QgsRectangle &extent, const
       conflictProfile = std::make_unique< QgsScopedRuntimeProfile >( QObject::tr( "Calculating conflicts" ), QStringLiteral( "rendering" ) );
     }
 
-    int idlp = 0;
```
Comment (nirvn) [inline src/core/pal/pal.cpp:399]: Thanks for cleaning up these obscure variable names.

File: src/core/pal/pal.h [hunk 7]
```diff
@@ -133,6 +134,17 @@ namespace pal
...
+       * This method:
+       *
+       * - preprocesses features, eg merging connected lines, chopping features at repeat distances
+       * - creates label candidates for every feature
+       * - purges candidates outside the map extent (respecting whether partial labels should be shown at the map boundary)
+       * - creates default fallback candidates for features with no valid candidates
+       * - collects obstacles
+       * - calculates candidate costs
+       * - calculates overlaps/conflicts
+       * - eliminates hard conflicts (forbidden placement)
```
Comment (nirvn) [inline src/core/pal/pal.h:147]: Forbidden! Nice nice, I'd change illegal to forbidden :)

File: src/core/labeling/rules/qgslabelingenginerule.h (hunk 2 above)
Comment (nyalldawson) [reply src/core/labeling/rules/qgslabelingenginerule.h:276]: Hmm... we have `QgsAbstractDatabaseProviderConnection::illegalFieldNames`, but nothing in current api with forbidden/prohibited/disallowed/banned. Is one other method enough to set the precendent?

File: src/core/labeling/rules/qgslabelingengineruleregistry.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/core/labeling/rules/qgslabelingengineruleregistry.cpp:26]: Ok, I've gone with:

- grey out the rule in the rules list
- show an explanatory tooltip
- don't allow users to create new versions of that rule

File: src/core/labeling/rules/qgslabelingenginerule.h (hunk 2 above)
Comment (nirvn) [reply src/core/labeling/rules/qgslabelingenginerule.h:286]: @nyalldawson , ideally we'd change illegalFieldNames to prohibited/forbiddenFileNames but that API is exposed to bindings so it's frozen right? 

Seems like settling for what preexisting keyword is a good argument enough then.

**Pull request merged**